ppa/archive/gale.py

```diff
@@ -0,0 +1,122 @@
+import logging
+import time
+
+import requests
+
+from django.conf import settings
+
+
+from ppa import __version__ as ppa_version
+
+
+logger = logging.getLogger(__name__)
+
+
+class GaleAPIError(Exception):
+    '''Base exception class for Gale API errors'''
+
+
+class GaleItemForbidden(GaleAPIError):
+    '''Permission denied to access item in Gale API'''
+
+
+class GaleItemNotFound(GaleAPIError):
+    '''Item not found in Gale API'''
+
+
+class GaleAPI:
+
+    #: base URL for all API requests
+    api_root = "https://api.gale.com/api"
+
+    instance = None
+
+    def __new__(cls):
+        # implement as a singleton
+        # adapted from https://softwareengineering.stackexchange.com/a/333710
+
+        # if no instance has been initialized, create and store on the class
+        if cls.instance is None:
+            cls.instance = super().__new__(cls)
+        # return the instance
+        return cls.instance
+
+    def __init__(self):
+        # NOTE: copied from hathi.py base api class; worth generalizing?
+
+        # create a request session, for request pooling
+        self.session = requests.Session()
+        # set a user-agent header, but  preserve requests version information
+        headers = {
+            "User-Agent": "ppa-django/%s (%s)"
+            % (ppa_version, self.session.headers["User-Agent"])
+        }
+        # include technical contact as From header, if set
+        tech_contact = getattr(settings, "TECHNICAL_CONTACT", None)
+        if tech_contact:
+            headers["From"] = tech_contact
+        self.session.headers.update(headers)
+        # error handling?
+        # username is needed to request an API key
+        self.username = settings.GALE_API_USERNAME
+
+    def _make_request(self, url, params=None, requires_api_key=True):
+        """Make a GET request with the configured session. Takes a url
+        relative to :attr:`api_root` and optional dictionary of parameters."""
+        # NOTE: also copied from hathi.py
+
+        # Returns the response for status 200 OK; raises
+        # :class:`HathiItemNotFound` for 404 and :class:`HathiItemForbidden`
+        # for 403.
+        # '''
+        url = "%s/%s" % (self.api_root, url)
+        rqst_opts = {}
+        if params:
+            rqst_opts["params"] = params
+
+        # add api key to parameters if neded for this request
+        if requires_api_key:
+            if "params" not in rqst_opts:
+                rqst_opts["params"] = {}
+            rqst_opts["params"]["api_key"] = self.api_key
+
+        start = time.time()
+        resp = self.session.get(url, **rqst_opts)
+        logger.debug("get %s %s: %f sec", url, resp.status_code, time.time() - start)
```

Comment:
looked up since I was curious — evidently you can do `resp.elapsed()` to get the request time, which is a `timedelta`, and `resp.elapsed().total_seconds()` for convenience. although if the response is huge, i'm not sure if the time to download/parse it is included

Reply:
oh, nice. I'll check how they compare

Reply:
Good instinct on the parsing, looks like you're right. I added logging so we could compare the two times for the same request, here's the logging for running the index script on the 6 items I'm currently testing with:

```
[09/Jun/2021 09:38:47] DEBUG:ppa.archive.gale::get https://api.gale.com/api/tools/generate_key 200: 0.181050 sec
[09/Jun/2021 09:38:47] DEBUG:ppa.archive.gale::get https://api.gale.com/api/tools/generate_key 200: 0.173248 sec [RESP elapsed]
[09/Jun/2021 09:39:15] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0119344579 200: 28.229499 sec
[09/Jun/2021 09:39:15] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0119344579 200: 23.927798 sec [RESP elapsed]
[09/Jun/2021 09:39:21] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0113954378 200: 5.633100 sec
[09/Jun/2021 09:39:21] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0113954378 200: 4.271696 sec [RESP elapsed]
[09/Jun/2021 09:39:26] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0111490063 200: 4.051841 sec
[09/Jun/2021 09:39:26] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0111490063 200: 3.067788 sec [RESP elapsed]
[09/Jun/2021 09:39:27] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0109750198 200: 1.174408 sec
[09/Jun/2021 09:39:27] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0109750198 200: 1.072275 sec [RESP elapsed]
[09/Jun/2021 09:39:33] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0103877546 200: 5.721335 sec
[09/Jun/2021 09:39:33] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0103877546 200: 4.063897 sec [RESP elapsed]
[09/Jun/2021 09:39:36] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0100160773 200: 3.074392 sec
[09/Jun/2021 09:39:36] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0100160773 200: 2.060866 sec [RESP elapsed]
```

I wonder if there's a way to make the json parsing more efficient (or at least less of a blocker) with a streaming response and iterative parsing?

Reply:
hmm, this is interesting. how does the 28sec response item compare to the average item? if there are a whole ton like it, i wonder if we will go over the 30min mark and need a new API key after all. i think there are a bunch of ways to come at this — streaming response/parse is one, but we could also maybe make the requests async, or do them synchronously but in parallel. on the other hand, if there won't ever be that many items, waiting another second or two for each request might not be that terrible.

Reply:
Wow! Just adding `stream=True` to the request get has a noticeable impact on the time:
```
[09/Jun/2021 09:48:52] DEBUG:ppa.archive.gale::get https://api.gale.com/api/tools/generate_key 200: 0.171706 sec
[09/Jun/2021 09:48:52] DEBUG:ppa.archive.gale::get https://api.gale.com/api/tools/generate_key 200: 0.164933 sec [RESP elapsed]
[09/Jun/2021 09:49:04] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0119344579 200: 11.703020 sec
[09/Jun/2021 09:49:04] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0119344579 200: 11.700790 sec [RESP elapsed]
[09/Jun/2021 09:49:12] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0113954378 200: 3.511412 sec
[09/Jun/2021 09:49:12] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0113954378 200: 3.509523 sec [RESP elapsed]
[09/Jun/2021 09:49:16] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0111490063 200: 2.294722 sec
[09/Jun/2021 09:49:16] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0111490063 200: 2.293044 sec [RESP elapsed]
[09/Jun/2021 09:49:18] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0109750198 200: 0.459720 sec
[09/Jun/2021 09:49:18] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0109750198 200: 0.457831 sec [RESP elapsed]
[09/Jun/2021 09:49:21] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0103877546 200: 3.087412 sec
[09/Jun/2021 09:49:21] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0103877546 200: 3.085139 sec [RESP elapsed]
[09/Jun/2021 09:49:25] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0100160773 200: 1.793068 sec
[09/Jun/2021 09:49:25] DEBUG:ppa.archive.gale::get https://api.gale.com/api/v1/item/GALE%7CCW0100160773 200: 1.791002 sec [RESP elapsed]
```

That puts the elapsed time close enough to actual time that we can just use that one and simplify the code, which is great.

Reply:
whoa!!! had no idea it was that easy. that's great!

Reply:
I've been wondering about making this script as a multiprocessing one with queues for import and indexing (which I think should ideally happen as part of import to avoid reindexing hathi content), but I'm doing my best to avoid premature optimization 😂